### PR TITLE
In#41/aws s3

### PR DESCRIPTION
**Pull request by michaeldotdev**
### What this PR does (required):
Implement client sends request to Express server for photo, then server grabs from S3 Bucket and send back to client to render. 

### Screenshots / Videos (required):
Screen Shot 2021-09-16 at 9 37 22 PM

### Any information needed to test this feature (required):
N/A

### Any issues with the current functionality (optional):

- Current image is hardcoded - not too sure on how to grab key params on client (have GET request code on client side?)
- Uploading from client then render photo to AvatarDisplay is incomplete

### CODE REVIEW

File: server/S3.js [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+const secretAccessKey = process.env.AWS_SECRET_ACCESS_KEY;
+
+const s3 = new S3({
+  region,
+  accessKeyId,
+  secretAccessKey,
+})
+
+// Function to handle uploading to S3 bucket
+function uploadPhoto(file) {
```
Comment (sundayezeilo) [inline server/S3.js:17]: Decorate with `async` keyword since it returns a `Promise`

File: server/app.js [hunk 2]
```diff
@@ -56,6 +60,23 @@ if (process.env.NODE_ENV === "production") {
...
+  const key = req.params.key;
+  console.log(key);
+  const downloadFromS3 = downloadPhoto(key);
+
+  downloadFromS3.pipe(res);
+})
+
+// POST request for uploading to Express server then pushing to S3 Bucket
+app.post('/profilePhotos', upload.single('profilePhoto'), async (req, res) => {
+  const file = req.file;
```
Comment (sundayezeilo) [inline server/app.js:75]: First, validate `file` in the request so that if it's missing, an error status should be returned to the client.

**Review by sundayezeilo — CHANGES_REQUESTED** [review]
You've done well on this one. :+1: However, the task is incomplete. You need to integrate this in the front-end.
You will receive support from our team on the complaint you left in the PR about grabbing the key params on the client. Meanwhile, address the comments I left below while you await our response.
Happy coding!

File: server/app.js (hunk 2 above)
Comment (michaeldotdev) [reply server/app.js:75]: I see, have a conditional if statement to check for whether or not the file is undefined?

File: server/S3.js (hunk 1 above)
Comment (michaeldotdev) [reply server/S3.js:17]: Where the function is call - it is async.

File: server/app.js (hunk 2 above)
Comment (rajivtitus) [reply server/app.js:75]: Yes, that is correct.

File: server/app.js [hunk 3]
```diff
@@ -56,6 +60,23 @@ if (process.env.NODE_ENV === "production") {
   });
 }
 
+// GET request from client > server > S3 bucket
+app.get('/profilePhotos/:key', (req, res) => {
+  console.log(res);
```
Comment (rajivtitus) [inline server/app.js:65]: Please remove console.log here and below

File: server/app.js [hunk 4]
```diff
@@ -56,6 +60,23 @@ if (process.env.NODE_ENV === "production") {
   });
 }
 
+// GET request from client > server > S3 bucket
+app.get('/profilePhotos/:key', (req, res) => {
```
Comment (rajivtitus) [inline server/app.js:64]: GET requests don't have a request body, you can send the key in the request body of a POST

**Comment by rajivtitus** [comment]
Closing these PRs as Michael is no longer a part of CAP :)

**Pull request closed without merging**